vortex-serde/src/layouts/write/metadata_accumulators.rs

```diff
@@ -0,0 +1,255 @@
+//! Metadata accumulators track the per-chunk-of-a-column metadata, layout locations, and row counts.
+
+use std::collections::VecDeque;
+use std::mem;
+use std::sync::Arc;
+
+use vortex::array::{BoolArray, NullArray, PrimitiveArray, StructArray, VarBinViewArray};
+use vortex::stats::{ArrayStatistics as _, Stat};
+use vortex::validity::Validity;
+use vortex::{Array, IntoArray as _};
+use vortex_buffer::{Buffer, BufferString};
+use vortex_dtype::{match_each_native_ptype, DType, NativePType, Nullability};
+use vortex_error::{vortex_bail, VortexError, VortexExpect as _, VortexResult};
+use vortex_scalar::Scalar;
+
+use super::layouts::Layout;
+use crate::stream_writer::ByteRange;
+
+pub fn new_metadata_accumulator(hint: usize, dtype: &DType) -> Box<dyn MetadataAccumulator> {
+    match dtype {
+        DType::Null => Box::new(ExtremaAccumulator::<()>::new(hint, into_null_array)),
+        DType::Bool(..) => Box::new(ExtremaAccumulator::<bool>::new(hint, into_bool_array)),
+        DType::Primitive(ptype, ..) => {
+            match_each_native_ptype!(ptype, |$P| {
+                Box::new(ExtremaAccumulator::<$P>::new(hint, into_primitive_array::<$P>))
+            })
+        }
+        DType::Utf8(..) => Box::new(ExtremaAccumulator::<BufferString>::new(
+            hint,
+            into_utf8_array,
+        )),
+        DType::Binary(..) => Box::new(ExtremaAccumulator::<Buffer>::new(hint, into_binary_array)),
+        DType::Struct(..) => Box::new(BasicAccumulator::new(hint)),
+        DType::List(..) => Box::new(BasicAccumulator::new(hint)),
+        DType::Extension(..) => Box::new(BasicAccumulator::new(hint)),
+    }
+}
+
+pub trait MetadataAccumulator {
+    fn push_chunk(&mut self, array: &Array) -> VortexResult<()>;
+
+    fn push_batch_byte_offsets(&mut self, batch_byte_offsets: Vec<u64>);
+
+    fn into_layouts_and_metadata(self: Box<Self>) -> VortexResult<(VecDeque<Layout>, StructArray)>;
```

Comment:
If you split this method into separate trait you can avoid passing the conversion function and instead specialize as necessary. 

```
pub trait IntoArrayMetadata {
    fn into_array_metadata(self: Box<Self>) -> VortexResult<(VecDeque<Layout>, StructArray)>;
}
```